### PR TITLE
More work on CA1859

### PR DESCRIPTION
**Pull request by geeknoid**
- Add support to control the API surface the rule reports on using the normal api_surface annotation in .editorconfig. The rule defaults to 'private' only behavior.

- Add support to analyze properties.

- Skip analyzing interface types since that ends up being completely useless as there's nothing to recommend. So this saves some cycles.

- Reject analyzing partial method implementations earlier to save some
cycles.

- Fixes https://github.com/dotnet/roslyn-analyzers/issues/6452

- Fixes https://github.com/dotnet/roslyn-analyzers/issues/6440

- Fixes https://github.com/dotnet/roslyn-analyzers/issues/6451

- No longer recommends changing the method signature (either return
types or parameter types) for virtual and override methods.

- Now properly recognizes interpolated string expressions,
nameof and sizeof, which all improve analysis accuracy.

### CODE REVIEW

**Comment by geeknoid** [comment]
@stephentoub Here's support for analyzing internal symbols.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs:126]: This should be documented in https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#supported-editorconfig-options

**Comment by mavasani** [comment]
I am not sure this will work as expected. Your analyzer is a SymbolStart/SymbolEnd analyzer, so if you have internal fields/methods whose uses are outside the containing type, then those won't be taken into account by analysis. If you want to analyze non-private members, you'll have to write a separate CompilationStart/CompilationEnd analyzer, most likely sharing majority of the code with your existing analyzer except for the analyzer action registration logic. You probably also need separate diagnostic IDs as compilation end analyzers are essentially build-only and won't run by default in live analysis and also cannot support code fixes.

**Comment by geeknoid** [comment]
@mavasani Hmmm. I see what you mean.

As it is, I think the "upgrade the return type of the method" analysis will work fine for public and internal methods, since this is just concerned with the implementation of a method. But the "upgrade fields" and "upgrade method parameter" analysis will not work since I don't have the right scope.

I'll give the CompilationStart/End thing a try. I guess I can use that strictly when asked to process internal/public symbols.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs
Comment (geeknoid) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs:138]: Done.

**Comment by geeknoid** [comment]
@mavasani Hi there, you folks need me to do anything else before this can be merged?

Thanks.

**Comment by buyaa-n** [comment]
@geeknoid this PR might have regression with overriden methods, I see bunch of new failures with overriden methods in runtime, [for example this](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/ArrayList.cs#L813)
```cs
    private sealed class IListWrapper : ArrayList
    {
           ...
          // Warns: Change return type of method 'Clone' from 'object' to 'System.Collections.ArrayList.IListWrapper' for improved performance
            public override object Clone() 
            {
                // This does not do a shallow copy of _list into a ArrayList!
                // This clones the IListWrapper, creating another wrapper class!
                return new IListWrapper(_list);
            }
          ....
    }
```
Some example of failures:
```log
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(1820,39) : Change return type of method 'GetRange' from 'System.Collections.ArrayList' to 'System.Collections.ArrayList.Range' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(2351,39) : Change return type of method 'GetRange' from 'System.Collections.ArrayList' to 'System.Collections.ArrayList.Range' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(1982,36) : Change return type of method 'Clone' from 'object' to 'System.Collections.ArrayList.ReadOnlyArrayList' forimproved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\Hashtable.cs(1331,36) : Change return type of method 'Clone' from 'object' to 'System.Collections.Hashtable' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(1727,36) : Change return type of method 'Clone' from 'object' to 'System.Collections.ArrayList.FixedSizeArrayList' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(2074,39) : Change return type of method 'GetRange' from 'System.Collections.ArrayList' to 'System.Collections.ArrayList.Range' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(1035,39) : Change return type of method 'GetRange' from 'System.Collections.ArrayList' to 'System.Collections.ArrayList.Range' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(854,41) : Change return type of method 'GetEnumerator' from 'System.Collections.IEnumerator' to 'System.Collections.ArrayList.IListWrapper.IListWrapperEnumWrapper' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(813,36) : Change return type of method 'Clone' from 'object' to 'System.Collections.ArrayList.IListWrapper' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(2262,36) : Change return type of method 'Clone' from 'object' to 'System.Collections.ArrayList.Range' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Collections\ArrayList.cs(1264,36) : Change return type of method 'Clone' from 'object' to 'System.Collections.ArrayList.SyncArrayList' for improved performance 
src\libraries\System.Private.CoreLib\src\System\Text\UTF7Encoding.cs(818,51) : Change return type of method 'CreateFallbackBuffer' from 'System.Text.DecoderFallbackBuffer' to 'System.Text.UTF7Encoding.DecoderUTF7FallbackBuffer' for improved performance 
src\coreclr\tools\Common\TypeSystem\Ecma\EcmaType.cs(57,22) : Change type of variable 'containingType' from 'Internal.TypeSystem.TypeDesc' to 'Internal.TypeSystem.DefType' for improved performance 
src\coreclr\tools\Common\TypeSystem\Ecma\SymbolReader\UnmanagedPdbSymbolReader.cs(746,40) : Change return type of method 'CreateObject' from 'object?' to 'Internal.TypeSystem.Ecma.UnmanagedPdbSymbolReader.SymUnmanagedDocumentWrapperCache.SymUnmanagedDocumentRcw?' for improved performance
src\coreclr\tools\Common\TypeSystem\Ecma\SymbolReader\UnmanagedPdbSymbolReader.cs(66,39) : Change return type of method 'CreateObject' from 'object' to 'Internal.TypeSystem.Ecma.UnmanagedPdbSymbolReader.ClrMetaHostWrapperCache.ClrMetaHostRcw' for improved performance 
src\coreclr\tools\Common\TypeSystem\Ecma\SymbolReader\UnmanagedPdbSymbolReader.cs(222,39) : Change return type of method 'CreateObject' from 'object' to 'Internal.TypeSystem.Ecma.UnmanagedPdbSymbolReader.MetaDataDispenserWrapperCache.MetaDataDispenserRcw' for improved performance 
src\coreclr\tools\Common\TypeSystem\Ecma\SymbolReader\UnmanagedPdbSymbolReader.cs(142,39) : Change return type of method 'CreateObject' from 'object' to 'Internal.TypeSystem.Ecma.UnmanagedPdbSymbolReader.CLRRuntimeInfoWrapperCache.ClrRuntimeInfoRcw' for improved performance 
src\coreclr\tools\Common\TypeSystem\Ecma\SymbolReader\UnmanagedPdbSymbolReader.cs(489,40) : Change return type of method 'CreateObject' from 'object?' to 'Internal.TypeSystem.Ecma.UnmanagedPdbSymbolReader.SymUnmanagedVariableWrapperCache.SymUnmanagedVariableRcw?' for improved performance 
src\coreclr\tools\Common\TypeSystem\Ecma\SymbolReader\UnmanagedPdbSymbolReader.cs(1070,39) : Change return type of method 'CreateObject' from 'object' to 'Internal.TypeSystem.Ecma.UnmanagedPdbSymbolReader.SymUnmanagedReaderWrapperCache.SymUnmanagedReaderRcw' for improved performance 
src\coreclr\tools\Common\TypeSystem\Ecma\SymbolReader\UnmanagedPdbSymbolReader.cs(376,40) : Change return type of method 'CreateObject' from 'object?' to 'Internal.TypeSystem.Ecma.UnmanagedPdbSymbolReader.SymUnmanagedNamespaceWrapperCache.SymUnmanagedNamespaceRcw?' for improved performance 
src\coreclr\tools\Common\TypeSystem\Ecma\SymbolReader\UnmanagedPdbSymbolReader.cs(857,40) : Change return type of method 'CreateObject' from 'object?' to 'Internal.TypeSystem.Ecma.UnmanagedPdbSymbolReader.SymUnmanagedMethodWrapperCache.SymUnmanagedMethodRcw?' for improved performance 
src\coreclr\tools\Common\TypeSystem\Ecma\SymbolReader\UnmanagedPdbSymbolReader.cs(572,40) : Change return type of method 'CreateObject' from 'object?' to 'Internal.TypeSystem.Ecma.UnmanagedPdbSymbolReader.SymUnmanagedScopeWrapperCache.SymUnmanagedScopeRcw?' for improved performance 
```
Other than these LGTM

**Review by mavasani — APPROVED** [review]

**Comment by mavasani** [comment]
I'll let @buyaa-n drive this PR and merge once she is ready to approve.

**Comment by geeknoid** [comment]
I was out for a bit, but I'm going to look into those false positives.

**Comment by geeknoid** [comment]
@buyaa-n The analyzer was recommending changing the signature of virtual methods and override methods which it really shouldn't have, just like it doesn't recommend changing the signature of methods that implement interfaces. I don't understand how come this wasn't a problem earlier. 

Anyway, all fixed now.

Thanks.

**Comment by buyaa-n** [comment]
Thank you, I see that is fixed and I don't see issue with the build with default settings.

When set the API surface to `internal, private` it is intermittently warning for overridden properties, like in below example it doesn't warn for `Assembly` property when `RuntimeTypeHandle.GetAssembly(this)` returns `RuntimeAssembly`, but warns for `BaseType` and `Module` properties, not sure what is the reason, it might be because `RuntimeTypeHandle.GetAssembly(this);` is public but others internal or private
```cs
    internal sealed partial class RuntimeType : TypeInfo, ICloneable
    {
        public override Assembly Assembly => RuntimeTypeHandle.GetAssembly(this); // not warn here
        public override Type? BaseType => GetBaseType(); // warns here
        public override bool IsByRefLike => RuntimeTypeHandle.IsByRefLike(this);
        public override bool IsGenericParameter => RuntimeTypeHandle.IsGenericVariable(this);
        public override bool IsTypeDefinition => RuntimeTypeHandle.IsTypeDefinition(this);
        public override bool IsSecurityCritical => true;
        public override bool IsSecuritySafeCritical => false;
        public override bool IsSecurityTransparent => false;
        public override MemberTypes MemberType => (IsPublic || IsNotPublic) ? MemberTypes.TypeInfo : MemberTypes.NestedType;
        public override int MetadataToken => RuntimeTypeHandle.GetToken(this); 
        public override Module Module => GetRuntimeModule(); // warns here
        public override Type? ReflectedType => DeclaringType;
        ...
    }
```
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/RuntimeType.cs#L17
```log
src\libraries\System.Private.CoreLib\src\System\RuntimeType.cs(17,31): warning CA1859: Change type of property 'BaseType' from 'System.Type?' to 'System.RuntimeType?' for improved performance
src\libraries\System.Private.CoreLib\src\System\RuntimeType.cs(26,32): warning CA1859: Change type of property 'Module' from 'System.Reflection.Module' to 'System.Reflection.RuntimeModule' for improved performance
```

**Review by buyaa-n — APPROVED** [review]
As this internal type is not flagged by default, I am OK with merging the PR as is and fix this issue later. Feel free to merge the PR, thanks you!

**Comment by geeknoid** [comment]
@buyaa-n I don't have a Merge button, so I don't think I have the permissions to merge. Could you do me the honors please?

**Comment by geeknoid** [comment]
@buyaa-n I fixed the issue with the overridden properties you found. Properties needed the same adjustment that methods did to not report overridden or virtual properties.

**Pull request merged**
